src/axom/quest/readers/C2CReader.cpp

```diff
@@ -377,5 +888,247 @@ void C2CReader::getLinearMesh(mint::UnstructuredMesh<mint::SINGLE_SHAPE>* mesh,
+    if(maxlen > len)
+    {
+      errPct = 100. * (1. - (len / maxlen));
+    }
+    return errPct;
+  };
+
+  constexpr size_t INITIAL_GUESS_NPTS = 100;
+  const double EPS_SQ = m_vertexWeldThreshold * m_vertexWeldThreshold;
+
+  // Clamp the lower error bound so it does not get impractically small.
+  percentError = axom::utilities::clampLower(percentError, 1.e-10);
+
+#ifdef AXOM_DEBUG_WRITE_ERROR_CURVES
+  // Make some curves for debugging.
+  FILE* ferr = fopen("error.curve", "wt");
+
+  FILE* fhcl = fopen("hicurvelen.curve", "wt");
+  fprintf(fhcl, "# hicurvelen\n");
+
+  FILE* fcl = fopen("curvelen.curve", "wt");
+  fprintf(fcl, "# curvelen\n");
+
+  FILE* fthresh = fopen("threshold.curve", "wt");
+  fprintf(fthresh, "# threshold\n");
+#endif
+  int contourCount = -1;
+
+  // Iterate over the contours and linearize each of them.
+  std::vector<Segment> S;
+  S.reserve(INITIAL_GUESS_NPTS);
+  for(const auto& nurbs : m_nurbsData)
+  {
+    NURBSInterpolator interpolator(nurbs, m_vertexWeldThreshold);
+    contourCount++;
+#ifdef AXOM_DEBUG_WRITE_ERROR_CURVES
+    fprintf(ferr, "# contour%d\n", contourCount);
+#endif
+
+    // Get the contour start/end parameters.
+    const double u0 = interpolator.startParameter(0);
+    const double u1 = interpolator.endParameter(interpolator.numSpans() - 1);
+
+    // Compute the start/end points of the whole curve.
+    const PointType p0 = interpolator.at(u0);
+    const PointType p1 = interpolator.at(u1);
+
+    // This segment represents the whole [0,1] interval.
+    Segment first;
+    first.u = u0;
+    first.point = p0;
+    first.length = sqrt(primal::squared_distance(p0, p1));
+    first.next_u = (u0 + u1) / 2.;
+    PointType next_point = interpolator.at(first.next_u);
+    first.next_length[0] = sqrt(primal::squared_distance(p0, next_point));
+    first.next_length[1] = sqrt(primal::squared_distance(next_point, p1));
+
+    // This segment just contains the end point of the interval.
+    Segment last;
+    last.u = u1;
+    last.point = p1;
+    last.length = 0.;
+    last.next_u = u1;
+    last.next_length[0] = 0.;
+    last.next_length[1] = 0.;
+
+    // Store the initial segments.
+    S.clear();
+    S.push_back(first);
+    S.push_back(last);
+
+    // If there is a single span in the contour then we already added its points.
+    if(interpolator.numSpans() > 1)
+    {
+      //---------------------------------------------------------------------
+      // Approximate the arc length of the whole curve by using a lot of
+      // line segments. Should we use quadrature instead here?
+      double hiCurveLen = 0.;
+      constexpr int MAX_NUMBER_OF_SAMPLES = 2000;
+      PointType prev = first.point;
+      for(int i = 1; i < MAX_NUMBER_OF_SAMPLES; i++)
+      {
+        double u = i / static_cast<double>(MAX_NUMBER_OF_SAMPLES - 1);
+        PointType cur = interpolator.at(u);
+        hiCurveLen += sqrt(primal::squared_distance(prev, cur));
+        prev = cur;
+      }
```

Comment:
Suggestion: Consider outsourcing the arc-length estimation to its own function.

Reply:
Done